### PR TITLE
cgen: fix struct field init with optional fixed array (fix #23195)

### LINKED ISSUE #23193: Struct optional field takes invalid value for fixed arrays
**V doctor:**

```
V full version: V 0.4.8 5ec9bb5.8835b6f
OS: linux, Linux version 6.6.8-arch1-1 (linux@archlinux) (gcc (GCC) 13.2.1 20230801, GNU ld (GNU Binutils) 2.41.0) #1 SMP PREEMPT_DYNAMIC Thu, 21 Dec 2023 19:01:01 +0000
Processor: 16 cpus, 64bit, little endian, 12th Gen Intel(R) Core(TM) i5-1240P

getwd: /home/ge/Code/v/netaddr/issues
vexe: /home/ge/.vlang/v
vexe mtime: 2024-12-17 15:55:29

vroot: OK, value: /home/ge/.vlang
VMODULES: OK, value: /home/ge/.vmodules
VTMP: OK, value: /tmp/v_1000

Git version: git version 2.47.1
Git vroot status: weekly.2024.50-63-g8835b6f2
.git/config present: true

CC version: cc (GCC) 14.2.1 20240910
emcc version: N/A
thirdparty/tcc status: thirdparty-linux-amd64 0134e9b9
```

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg struct_optional_fixed_array_field.vv`

```v
type Arr = [4]u8

struct Foo {
	bar int
	baz ?Arr
}

fn main() {
	arr := Arr([u8(5), 4, 3, 2]!)
	f := Foo{
		bar: 1
		baz: arr // Option(Arr([5, 4, 3, 2])) expected, but Option(none) printed
	}
	println(f)
	assert f.bar == 1
	assert f.baz as Arr == [u8(5), 4, 3, 2]!
}
```

**What did you expect to see?**

Assertions passed.

The same code works for non-fixed arrays: https://play.vlang.io/p/7e5f0012ed

**What did you see instead?**

```
Foo{
    bar: 1
    baz: Option(none)
}
struct_optional_fixed_array_field.vv:16: FAIL: fn main.main: assert f.baz as main.Arr == [u8(5), 4, 3, 2]u8{}
  left value: f.baz as main.Arr = Arr([2, 0, 0, 0])
  right value: [u8(5), 4, 3, 2]u8{} = [5, 4, 3, 2]
V panic: Assertion failed...
v hash: 8835b6f
/tmp/v_1000/struct_optional_fixed_array_field.01JFAP91B2BNTK1GK8VHY1H00M.tmp.c:4058: at _v_panic: Backtrace
/tmp/v_1000/struct_optional_fixed_array_field.01JFAP91B2BNTK1GK8VHY1H00M.tmp.c:5956: by main__main
/tmp/v_1000/struct_optional_fixed_array_field.01JFAP91B2BNTK1GK8VHY1H00M.tmp.c:5982: by main
```

Also test fails for unaliased array:

```v
struct Foo {
	bar int
	baz ?[4]u8
}

fn main() {
	arr := [u8(5), 4, 3, 2]!
	f := Foo{
		bar: 1
		baz: arr
	}
	println(f)
	assert f.bar == 1
	assert f.baz as [4]u8 == [u8(5), 4, 3, 2]!
}
```

> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

<sub><a href="https://huly.app/guest/vlang-66f40c4d-a476b54c67-771fdd?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJsaW5rSWQiOiI2NzYxYTRiMWYxNDRmNTg1YWU0ZGNjMGEiLCJndWVzdCI6InRydWUiLCJlbWFpbCI6IiNndWVzdEBoYy5lbmdpbmVlcmluZyIsIndvcmtzcGFjZSI6InctYWxleGFuZGVyLXZsYW5nLTY2ZjQwYzRkLWE0NzZiNTRjNjctNzcxZmRjIn0.WJHjyyr3VsGXUNa7zqsxQThVZI72iCs45MdoKXPpwCU">Huly&reg;: <b>V_0.6-21630</b></a></sub>

### LINKED ISSUE #23195: C error when assigning fixed array alias to struct optional field
**V doctor:**

```
V full version: V 0.4.8 5ec9bb5.8835b6f
OS: linux, Linux version 6.6.8-arch1-1 (linux@archlinux) (gcc (GCC) 13.2.1 20230801, GNU ld (GNU Binutils) 2.41.0) #1 SMP PREEMPT_DYNAMIC Thu, 21 Dec 2023 19:01:01 +0000
Processor: 16 cpus, 64bit, little endian, 12th Gen Intel(R) Core(TM) i5-1240P

getwd: /home/ge/Code/v/netaddr/issues
vexe: /home/ge/.vlang/v
vexe mtime: 2024-12-17 15:55:29

vroot: OK, value: /home/ge/.vlang
VMODULES: OK, value: /home/ge/.vmodules
VTMP: OK, value: /tmp/v_1000

Git version: git version 2.47.1
Git vroot status: weekly.2024.51-9-g8835b6f2
.git/config present: true

CC version: cc (GCC) 14.2.1 20240910
emcc version: N/A
thirdparty/tcc status: thirdparty-linux-amd64 0134e9b9
```

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg struct_optional_aliased_fixed_array_field.vv`

```v
type Arr = [4]u8

struct Foo {
	bar int
	baz ?Arr
}

fn main() {
	f := Foo{
		bar: 1
		baz: Arr([u8(5), 4, 3, 2]!)
	}
	println(f)
	assert f.baz as Arr == [u8(5), 4, 3, 2]!
}
```

**What did you expect to see?**

Assertion passed, no C errors.

**What did you see instead?**

```
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/struct_optional_aliased_fixed_array_field.01JFAPX70SQ3AVKPSXNNJCDG8G.tmp.c:5919: error: invalid operand types for binary operation
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

<sub><a href="https://huly.app/guest/vlang-66f40c4d-a476b54c67-771fdd?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJsaW5rSWQiOiI2NzYxYTYyMWYxNDRmNTg1YWU0ZGNjMmUiLCJndWVzdCI6InRydWUiLCJlbWFpbCI6IiNndWVzdEBoYy5lbmdpbmVlcmluZyIsIndvcmtzcGFjZSI6InctYWxleGFuZGVyLXZsYW5nLTY2ZjQwYzRkLWE0NzZiNTRjNjctNzcxZmRjIn0.mvPYC8_YZwO_0-uS-NBKyOtns7f0MFjAItFrU-PYPFE">Huly&reg;: <b>V_0.6-21632</b></a></sub>

### PR DESCRIPTION
**Pull request by yuyi98**
This PR fix struct field init with optional fixed array (fix #23195, fix #23193).

- Fix struct field init with optional fixed array.
- Add test.

```v
type Arr = [4]u8

struct Foo {
	bar int
	baz ?Arr
}

fn main() {
	f1 := Foo{
		bar: 1
		baz: Arr([u8(5), 4, 3, 2]!)
	}
	println(f1)
	assert f1.baz as Arr == [u8(5), 4, 3, 2]!

	f2 := Foo{
		bar: 1
		baz: ?Arr(none)
	}
	println(f2)
	assert f2.bar == 1
	assert f2.baz == none

	arr := Arr([u8(5), 4, 3, 2]!)
	f3 := Foo{
		bar: 1
		baz: arr
	}
	println(f3)
	assert f3.bar == 1
	assert f3.baz as Arr == [u8(5), 4, 3, 2]!
}

PS D:\Test\v\tt1> v run .
Foo{
    bar: 1
    baz: Option(    Arr([5, 4, 3, 2]))
}
Foo{
    bar: 1
    baz: Option(none)
}
Foo{
    bar: 1
    baz: Option(    Arr([5, 4, 3, 2]))
}
```

<sub><a href="https://huly.app/guest/vlang-66f40c4d-a476b54c67-771fdd?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJsaW5rSWQiOiI2NzYyNTdjMGYxNDRmNTg1YWU0ZGRkOWYiLCJndWVzdCI6InRydWUiLCJlbWFpbCI6IiNndWVzdEBoYy5lbmdpbmVlcmluZyIsIndvcmtzcGFjZSI6InctYWxleGFuZGVyLXZsYW5nLTY2ZjQwYzRkLWE0NzZiNTRjNjctNzcxZmRjIn0.IRF-FLV5vcI4xBqru46dOyW5d_jmH6F4JSPwPbH0S5k">Huly&reg;: <b>V_0.6-21636</b></a></sub>

### CODE REVIEW

File: vlib/v/tests/structs/struct_field_init_with_fixed_array_opt_test.v [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+struct Foo {
+	bar int
+	baz ?Arr
+}
+
+fn test_struct_field_init_with_fixed_array_opt() {
+	f := Foo{
+		bar: 1
+		baz: Arr([u8(5), 4, 3, 2]!)
```
Comment (felipensp) [inline vlib/v/tests/structs/struct_field_init_with_fixed_array_opt_test.v:11]: It would be good add a test for initializing it with `baz: ?Arr(none)`

File: vlib/v/gen/c/assign.v [hunk 2]
```diff
@@ -157,6 +157,18 @@ fn (mut g Gen) expr_with_opt(expr ast.Expr, expr_typ ast.Type, ret_typ ast.Type)
 	return ''
 }
 
+fn (mut g Gen) expr_with_fixed_array_opt(expr ast.Expr, expr_typ ast.Type, ret_typ ast.Type) {
+	tmp_var := g.new_tmp_var()
+	stmt_str := g.go_before_last_stmt().trim_space()
+	styp := g.styp(expr_typ)
+	g.empty_line = true
+	g.writeln('${styp} ${tmp_var} = {.state = 0};')
```
Comment (felipensp) [inline vlib/v/gen/c/assign.v:165]: What if expr has a none one? `state` should not be 0 in this case.

Comment (yuyi98) [reply vlib/v/gen/c/assign.v:165]: none does not belong to this function call.

**Comment by felipensp** [comment]
Excellent, @yuyi98! Thanks.
